# Convergence of Lück approximation

We investigate the convergence of Lück's approximation theorem by computing the von Neumann rank of a "naturally occurring" matrix.\
We choose the second differential of a chain complex for the universal cover of the manifold v1539(5,1).

In [1]:
import twisted_l2
import regina
import snappy

## Compute the chain complex of the universal cover:

In [2]:
twisted_l2.load_hap()

true

In [3]:
mfd = snappy.Manifold("v1539(5,1)")
tri = regina.Triangulation3(mfd.filled_triangulation())

In [4]:
fl = twisted_l2.regina_tri_to_face_lattice(tri, ideal=False)

In [5]:
cw = twisted_l2.cw_complex(fl)

In [6]:
number_of_cells = twisted_l2.gap_member(cw, "nrCells")
print("Number of cells in each dimension:")
for j in range(4):
    print(f"{j}: {number_of_cells(j)}")

Number of cells in each dimension:
0: 98
1: 674
2: 1152
3: 576


In [7]:
%time chain_complex = twisted_l2.equivariant_cc(cw, gap=False)

CPU times: user 113 ms, sys: 307 µs, total: 114 ms
Wall time: 113 ms


In [8]:
[chain_complex.dimension(i) for i in range(5)]

[1, 3, 3, 1, 0]

In [9]:
G = twisted_l2.get_fundamental_group(chain_complex)

In [10]:
cc = twisted_l2.get_differentials(chain_complex)

The variable `chain_complex` wraps a GAP object,
while `cc` is simply a list of matrices over the group algebra of `G`. \
Actually, for technical reasons, we use the free group on the generators of `G` instead of `G` itself.

The above cell is equivalent to
```
cc = [twisted_l2.boundary_operator(chain_complex, i, G) for i in range(1, 4)]
```

## Or load precomputed data:

We can also save/load the group and chain complex.\
This is useful to deal with non-deterministic results from CW complex simplifications etc.

In [2]:
# twisted_l2.save_to_file("data/v1539-5-1-3x3-2.json", group=G, cc=cc)
G, cc = twisted_l2.load_from_file("data/v1539-5-1-3x3.json")

In [3]:
G

Finitely presented group < x0, x1, x2 | (x0^2*x2^-5*x1^-3)^4*x0^2*x1^2, x0^2*x2^-5*x1^-2*x2, x2^9*x0^-3*x1^3*x2^-3 >

In [4]:
# Choose the desired configuration (see "configs.py")
twisted_l2.configs.LogOptions.LEVEL = twisted_l2.INFO

## Now we compute the rank:

In [5]:
A = cc[1]

This is a $3\times 3$ matrix.

In [6]:
%time twisted_l2.von_neumann_rank(G, A, (1,))

Size of Fin: 4
|L| = 4
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (12, 12)
Rank: 1.75 (rounds up to 2)
CPU times: user 26.3 ms, sys: 3.27 ms, total: 29.6 ms
Wall time: 28.8 ms


7/4

In [7]:
%time twisted_l2.von_neumann_rank(G, A, (2,))

Size of Fin: 32
|L| = 32
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (96, 96)
Rank: 1.96875 (rounds up to 2)
CPU times: user 68 ms, sys: 9.06 ms, total: 77.1 ms
Wall time: 76.3 ms


63/32

In [8]:
%time twisted_l2.von_neumann_rank(G, A, (0,1))

Size of Fin: 9
|L| = 9
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (27, 27)
Rank: 1.888888888888889 (rounds up to 2)
CPU times: user 11.1 ms, sys: 2.96 ms, total: 14.1 ms
Wall time: 14.2 ms


17/9

In [9]:
%time twisted_l2.von_neumann_rank(G, A, (1,1))

Size of Fin: 36
|L| = 36
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (108, 108)
Rank: 1.861111111111111 (rounds up to 2)
CPU times: user 25.6 ms, sys: 5.92 ms, total: 31.5 ms
Wall time: 31 ms


67/36

In [10]:
%time twisted_l2.von_neumann_rank(G, A, (0,2))

Size of Fin: 243
|L| = 243
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (729, 729)
Rank: 1.995884773662551 (rounds up to 2)
CPU times: user 509 ms, sys: 3.58 ms, total: 512 ms
Wall time: 512 ms


485/243

In [11]:
%time twisted_l2.von_neumann_rank(G, A, (0,0,1))

Size of Fin: 25
|L| = 25
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (75, 75)
Rank: 1.64 (rounds up to 2)
CPU times: user 25.7 ms, sys: 3.59 ms, total: 29.3 ms
Wall time: 28.8 ms


41/25

In [12]:
%time twisted_l2.von_neumann_rank(G, A, (0,0,2))

Size of Fin: 3125
|L| = 3125
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (9375, 9375)
Rank: 1.99072 (rounds up to 2)
CPU times: user 1min 42s, sys: 187 ms, total: 1min 42s
Wall time: 1min 42s


6221/3125

In [13]:
%time twisted_l2.von_neumann_rank(G, A, (0,0,0,1))

Size of Fin: 49
|L| = 49
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (147, 147)
Rank: 1.979591836734694 (rounds up to 2)
CPU times: user 29.8 ms, sys: 3.31 ms, total: 33.1 ms
Wall time: 32.8 ms


97/49

In [14]:
%time twisted_l2.von_neumann_rank(G, A, (0,0,0,0,1))

Size of Fin: 121
|L| = 121
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (363, 363)
Rank: 1.991735537190083 (rounds up to 2)
CPU times: user 129 ms, sys: 11 µs, total: 129 ms
Wall time: 129 ms


241/121

In [15]:
%time twisted_l2.von_neumann_rank(G, A, (0,0,0,0,0,1))

Size of Fin: 169
|L| = 169
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (507, 507)
Rank: 1.994082840236686 (rounds up to 2)
CPU times: user 213 ms, sys: 6.62 ms, total: 219 ms
Wall time: 219 ms


337/169

In [16]:
%time twisted_l2.von_neumann_rank(G, A, (0,0,0,0,0,0,1))

Size of Fin: 289
|L| = 289
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (867, 867)
Rank: 1.996539792387543 (rounds up to 2)
CPU times: user 627 ms, sys: 3.32 ms, total: 631 ms
Wall time: 631 ms


577/289

In [17]:
%time twisted_l2.von_neumann_rank(G, A, (1,1,1))

Size of Fin: 900
|L| = 900
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (2700, 2700)
Rank: 1.985555555555556 (rounds up to 2)
CPU times: user 5.58 s, sys: 16 µs, total: 5.58 s
Wall time: 5.59 s


1787/900

In [18]:
%time twisted_l2.von_neumann_rank(G, A, (0,1,0,1))

Size of Fin: 441
|L| = 441
Constructing matrix over Q[L]...
Computing rank...
Dimensions of N: (1323, 1323)
Rank: 1.997732426303855 (rounds up to 2)
CPU times: user 1.29 s, sys: 3.31 ms, total: 1.29 s
Wall time: 1.29 s


881/441

| `exps`         | $|L|$ | Rank   | Time (ms) | $|L| \cdot$ error |
| :---           | ---:  | :---   | ---:      | ---: |
|(1)             | 4    | 1.75    | 29     | 1  | 
|(2)             | 32   | 1.96875 | 76     | 1  |
|(0,1)           | 9    | 1.88889 | 14     | 1  |
|(1,1)           | 36   | 1.86111 | 31     | 5  |
|(0,2)           | 243  | 1.99588 | 512    | 1  |
|(0,0,1)         | 25   | 1.64    | 29     | 9  |
|(0,0,2)         | 3125 | 1.99072 | 102000 | 29 |
|(0,0,0,1)       | 49   | 1.97959 | 33     | 1  |
|(0,0,0,0,1)     | 121  | 1.99174 | 129    | 1  |
|(0,0,0,0,0,1)   | 169  | 1.99408 | 219    | 1  |
|(0,0,0,0,0,0,1) | 289  | 1.99654 | 631    | 1  |
|(1,1,1)         | 900  | 1.98556 | 5590   | 13 |
|(0,1,0,1)       | 441  | 1.99773 | 1290   | 1  |

Since the chain complex is $L^2$-acyclic and has dimensions $(1,3,3,1)$, we know that the true value of the rank is $2$.

In general, the error is inversely proportional to the size of the finite quotient.\
However, the coefficient worsens when the quotient involves a $5$-group or when it is a product of several $p$-groups.\
It is advisable to try out class-$1$ $p$-quotients to find _special_ primes (in this case, $5$) quickly,
and avoid them unless absolutely necessary.

The running time depends strongly on $|L|$, due to the rank computation, with exponent $$\frac{\log(102000/5590)}{\log(3125/900)}.$$

In [20]:
(log(102000/5590) / log(3125/900)).n()

2.33290943118416

This value is consistent with the superquadratic complexity of matrix rank.